mono/mini/mini-windows.c

```diff
@@ -152,6 +152,42 @@ get_win32_restore_stack (void)
 
 	return start;
 }
+#elif HAVE_API_SUPPORT_WIN32_RESET_STKOFLW && TARGET_ARM64
+static gpointer
+get_win32_restore_stack(void)
+{
+	static guint8* start = NULL;
+	guint8* code;
+
+	if (start)
+		return start;
+
+	const int size = 64;
+
+	/* restore_stack (void) */
+	start = code = mono_global_codeman_reserve(size);
+
+	arm_stpx_pre(code, ARMREG_FP, ARMREG_LR, ARMREG_SP, -0x10);
+	arm_movx(code, ARMREG_FP, ARMREG_SP);
+	arm_adrpx(code, ARMREG_R8, _resetstkoflw);
+	arm_addx(code, ARMREG_R8, ARMREG_R8, ((int)_resetstkoflw) & 0xFFF);
```

Comment:
What does the mask/add do?

Reply:
((int)_resetstkoflw) & 0xFFF is to not hard code the offset, if the function position moved when the code was re-compiled

Reply:
The call thats needs to be made is ADRL, however the macro for this doesnt exist, instead its split into two ADRP and ADD. 
((int)_resetstkoflw) & 0xFFF is to not hard code the offset, if the function position moved when the code was re-compiled

Reply:
This needs a good comment explaining it. I am trying to understand this logic as well. This is what I got so far:

In ARM64, each instruction is strictly 4 bytes in length. Which means that you cannot load large immediate values (such as function addresses) as they simply cannot be encoded into the instruction. When it comes to function addresses, this causes complications. Since you cannot encode absolute addresses, all function call instructions on ARM64 either take the address from the register (like on the line below), or use PC-relative addressing, where the relative address is limited to 11 bits (I think?), which gives you +-1 MB jump range.

Since we cannot guarantee that `_resetstkoflw` will be located within 1 MB of wherever JITted code is, we need to call the function using the absolute address which is taken from a register. Now, loading an absolute address to a register has the same problem: you don't have enough bits. So you need to load it using relative addressing. And since even relative addressing is limited, you do it in two instructions: ADRP (https://developer.arm.com/documentation/dui0801/h/A64-General-Instructions/ADRP) which loads the high 52 bits (`64 - 12 = 52`) and then the ADD instruction takes care of the low 12-bits, that's why Jem masked the _resetstkoflw label with 0xFFF.

There are two things that I think *might* be going wrong, but I can't tell for sure:

1. The resulting address must be relative to the PC register. ADRP instruction takes care of that for you:

    #define arm_format_adrp(p, op, rd, target) do { guint64 imm1 = (guint64)(target); guint64 imm2 = (guint64)(p); int _imm = imm1 - imm2; arm_emit ((p), ((op) << 31) | (((_imm) & 0x3) << 29) | (0x10 << 24) | (((_imm >> 2) & 0x7ffff) << 5) | ((rd) << 0)); } while (0)

However, I don't think `arm_addx` does that so you might need to calculate relative address manually before masking the address.

2. I don't know about this for sure, but I think you should cast `_resetstkoflw` to `uintptr_t` instead of `int` to prevent accidental overflows. I don't know what downcasting overflow behavior is on ARM64 off the top of my head.

Reply:
I actually must be confusing myself too... The resulting address is going to be absolute. The address supplied to the ADRP instruction needs to be relative, and this instruction will put the absolute address to the destination register. So adding to the register needs to result in absolute address, so looks this code is actually correct! Sorry, this is quite a bit confusing :).